lib/hobbes/boot/dependent.hob

```diff
@@ -0,0 +1,3 @@
+
+type Type = ^x.|Prim:([char] * (() + x)), Record:[[char]*x], Variant:[[char]*x*int], OpaquePtr:([char]*int*bool), Exists:([char]*x), Recursive:([char]*x), TAbs:([[char]]*x), TApp:(x*[x]), Func:([x]*x), Array:x, FixedArray:(x*x)|
```

Comment:
This is awesome, nicely simplifies the crazy ad-hoc C++ encoding of the same data (and embeds in hobbes perfectly).

Reply:
Sort-of, note we are missing TExpr...

Reply:
Could we hide the TExpr's Expr behind an existential?

Reply:
I missed out TVar because I didn't want them as input, but it might actually be useful to be able to construct them.

Reply:
I guess that in order to go fully into dependent types, we just need `Expr` (with a few more constant forms for e.g. the unit type, the universe type hierarchy, ...) since we would no longer differentiate between expressions and types (or we'd say an expression belongs to a "type" which is another expression which belongs to a "type" etc until we terminate in a universe, where it'll always be the case that a universe belongs to the type of its successor universe).

Reply:
I think I can add this as TExpr:<hobbes. ExprPtr>, then give another unqualified to reify that pointer.

Reply:
Making the internal representations of types and expressions into these bindable recursive variants seems like the right way to go.  Then all of that boilerplate code to go between the two equivalent representations can be eliminated.

After that, reducing the type/expr types to just expr will get us to dependent types.

Reply:
Yep. Removing boilerplate is always the way to go.

Reply:
Now if we also constructed an Expr type, allowed that to be returned and added a SymDef that reifies that, we could write a generic apply.